newIDE/app/src/EventsSheet/index.js

```diff
@@ -1105,85 +1126,106 @@ export default class EventsSheet extends React.Component<Props, State> {
+                                  }
                                 ),
-                            },
-                            {
-                              label: 'Add Sub Event',
-                              click: () => this.addSubEvents(),
-                              enabled: this._selectionCanHaveSubEvents(),
-                            },
-                            {
-                              label: 'Add Other',
-                              submenu: this.state.allEventsMetadata.map(
-                                metadata => {
-                                  return {
-                                    label: metadata.fullName,
-                                    click: () =>
-                                      this.addNewEvent(metadata.type),
-                                  };
+                              },
+                              { type: 'separator' },
+                              {
+                                label: 'Undo',
+                                click: this.undo,
+                                enabled: canUndo(this.state.history),
+                                accelerator: 'CmdOrCtrl+Z',
+                              },
+                              {
+                                label: 'Redo',
+                                click: this.redo,
+                                enabled: canRedo(this.state.history),
+                                accelerator: 'CmdOrCtrl+Shift+Z',
+                              },
+                              { type: 'separator' },
+                              {
+                                label: 'Extract Events to a Function',
+                                click: () => this.extractEventsToFunction(),
+                              },
+                              {
+                                label: 'Move Events into a Group',
+                                click: () => this.moveEventsIntoNewGroup(),
+                              },
+                              {
+                                label: 'Analyze Objects Used in this Event',
+                                click: this._openEventsContextAnalyzer,
+                              },
+                            ];
+
+                            const edition = {
+                              label: 'Edit',
+                              click: () => this.openEventTextDialog(),
+                            };
+
+                            getSelectedEvents(this.state.selection).forEach(
+                              event => {
+                                if (
+                                  event.getType() ===
+                                    'BuiltinCommonInstructions::Comment' ||
+                                  event.getType() ===
+                                    'BuiltinCommonInstructions::Group'
+                                ) {
+                                  contextList.unshift(edition);
```

Comment:
Rather than unshift-ing, let's try to keep a single array (so that it's more readable - I prefer to avoid "complex logic" in the render of components) (and also, this `render` is super long already).
Use `visible: ` set to true or false (next to label, and click) when defining the menu item.

Also careful about the forEach and unshift here, if somehow you have multiple comment or group selected, you'll have an "Edit" menu item inserted for each of them - which is probably not what you want ;)

Instead make a function for it to pass to `visible`:

`visible: filterEditableWithEventTextDialog(getSelectedEvents(this.state.selection)).length > 0`

and make `filterEditableWithEventTextDialog` a function:

```
const filterEditableWithEventTextDialog = (events: Array<gdBaseEvent>) => {
  return events.filter(event => 
    ['BuiltinCommonInstructions::Group', 'BuiltinCommonInstructions::Comment'].includes(event.getType())
  );
}
```

(not tested but you get the idea).
Last thing, where to put this function? It contains knowledge related to `EventTextDialog` so `EventTextDialog.js` file is a good place. Put it here, export it (`export const canBeEditedWithEventTextDialog = ...`).

And finally import it in EventsSheet.

This will allow to have just an array for the menu items in EventsSheet, and the "complex logic" colocated with the rest of the source code of EventTextDialog. It's better because now you only have a single file that contains all the rules telling which events are supported by EventTextDialog - rather than two (so that you have less change to forget to update one if you add a new event in the future).